src/charm.py

```diff
@@ -3,115 +3,267 @@
 
-        self._reconcile_runners()
-        event.set_results({"status": "Completed runner check"})
+    def _on_flush_runners_action(self, event):
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
+            event.fail("Missing token or org/repo path config")
+            return
 
-    def _reconcile_runners(self):
-        delta = self.config["quantity"] - self._runner.active_count()
-        logger.info(f"Reconciling {delta} runners")
-        self._runner.remove_runners()
-        while delta > 0:
-            self.unit.status = MaintenanceStatus(f"Installing {delta} runner(s)")
-            try:
-                self._runner.create(image="ubuntu", virt=self.config["virt-type"])
-            except RuntimeError as e:
-                logger.error("Failed to create runner")
-                logger.error(f"Error: {e}")
-            delta = self.config["quantity"] - self._runner.active_count()
-        self.unit.status = ActiveStatus("Active and registered")
+        try:
+            runner_manager.clear()
+            delta = runner_manager.reconcile(self.config["quantity"])
+        except Exception as e:
+            logger.exception("Failed to flush runners")
+            event.fail(f"Failed to flush runners: {e}")
+            return
+        self._on_check_runners_action(event)
+        event.set_results({"delta": delta})
 
     def _on_stop(self, event):
-        self._remove_cron("check-runners")
-        self._runner.remove_runners()
-
-    def _add_cron(self, event, interval):
-        """Add a cron job for the provided event to run at the provided interval."""
-        root_cron = self._cron_tab(user="root")
-        comment = f"Charm cron for {event}"
-
-        # Remove if it exists and write with the possibly updated interval
-        self._remove_cron(event)
-
-        unit = self.unit.name
-        dispatch = str(self.charm_dir / "dispatch")
-        command = f"juju-run {unit} 'JUJU_DISPATCH_PATH={event} {dispatch}'"
-        job = root_cron.new(command=command, comment=comment)
-        job.setall(interval)
-        root_cron.write()
-
-    def _remove_cron(self, event):
-        """Remove cron job for provided event."""
-        root_cron = self._cron_tab(user="root")
-        try:
-            job = next(root_cron.find_comment(f"Charm cron for {event}"))
-            root_cron.remove(job)
-            root_cron.write()
-        except StopIteration:
-            pass  # Not found
+        self._disable_event_timer("update-runner-bin")
+        self._disable_event_timer("reconcile-runners")
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
+            try:
+                runner_manager.clear()
+            except Exception:
+                # Log but ignore error since we're stopping anyway.
+                logger.exception("Failed to clear runners")
+
+    def _render_event_tmpl(self, tmpl_type, event_name, context):
```

Comment:
Minor nit, but this seems better named as `_write_event_tmpl`. Offhand, I was surprised that `self._render_event_tmpl(...)` didn't return a rendered template